bindings/rust/evmc-vm/src/lib.rs

```diff
@@ -2,6 +2,195 @@ pub extern crate evmc_sys;
 pub use evmc_sys as ffi;
 
 // TODO: Add helpers for host interface
-// TODO: Add convenient helpers for evmc_result
 // TODO: Add convenient helpers for evmc_execute
 // TODO: Add a derive macro here for creating evmc_create
+
+/// EVMC result structure.
+pub struct ExecutionResult {
+    status_code: ffi::evmc_status_code,
+    gas_left: i64,
+    output: Option<Vec<u8>>,
+    create_address: ffi::evmc_address,
+}
+
+impl ExecutionResult {
+    pub fn new(
+        _status_code: ffi::evmc_status_code,
+        _gas_left: i64,
+        _output: Option<Vec<u8>>,
+        _create_address: ffi::evmc_address,
+    ) -> Self {
+        ExecutionResult {
+            status_code: _status_code,
+            gas_left: _gas_left,
+            output: _output,
+            create_address: _create_address,
+        }
+    }
+
+    pub fn get_status_code(&self) -> ffi::evmc_status_code {
+        self.status_code
+    }
+
+    pub fn get_gas_left(&self) -> i64 {
+        self.gas_left
+    }
+
+    pub fn get_output(&self) -> Option<&Vec<u8>> {
+        self.output.as_ref()
+    }
+
+    pub fn get_create_address(&self) -> Option<&ffi::evmc_address> {
+        // Only return Some if the address is valid (e.g. the status is EVMC_SUCCESS)
+        if self.status_code == ffi::evmc_status_code::EVMC_SUCCESS {
+            Some(&self.create_address)
+        } else {
+            None
+        }
+    }
+}
+
+impl From<ffi::evmc_result> for ExecutionResult {
+    fn from(result: ffi::evmc_result) -> Self {
+        let ret = ExecutionResult {
+            status_code: result.status_code,
+            gas_left: result.gas_left,
+            output: if !result.output_data.is_null() {
+                // Pre-allocate a vector.
+                let mut buf: Vec<u8> = Vec::with_capacity(result.output_size);
+
+                unsafe {
+                    // Set the len of the vec manually.
+                    buf.set_len(result.output_size);
+                    // Copy from the C struct's buffer to the vec's buffer.
+                    std::ptr::copy(result.output_data, buf.as_mut_ptr(), result.output_size);
+                }
+
+                Some(buf)
+            } else {
+                None
+            },
+            create_address: result.create_address,
+        };
+
+        // Release allocated ffi struct.
+        unsafe {
+            if result.release.is_some() {
+                result.release.unwrap()(&result as *const ffi::evmc_result);
```

Comment:
This is nitpicking, but could format this as
```
if ... {
  unsafe { release ..
```

I'm fine either way.

Reply:
I think rustfmt will complain.

Reply:
Why? The proposal is that only the release call is unsafe and not the condition. Either way it doesn't matter.

Reply:
oh, i thought this was mentioning the indentation. will do that